compiler/rustc_target/src/spec/mod.rs

```diff
@@ -320,8 +377,24 @@ impl LinkerFlavor {
     }
 
     pub fn check_compatibility(self, cli: LinkerFlavorCli) -> Option<String> {
-        // The CLI flavor should be compatible with the target if it survives this roundtrip.
-        let compatible = |cli| cli == self.with_cli_hints(cli).to_cli();
+        let compatible = |cli| {
+            // The CLI flavor should be compatible with the target if:
+            // 1. they are counterparts: they have the same principal flavor.
+            match (self, cli) {
+                (LinkerFlavor::Gnu(..), LinkerFlavorCli::Gnu(..))
+                | (LinkerFlavor::Darwin(..), LinkerFlavorCli::Darwin(..))
+                | (LinkerFlavor::WasmLld(..), LinkerFlavorCli::WasmLld(..))
+                | (LinkerFlavor::Unix(..), LinkerFlavorCli::Unix(..))
+                | (LinkerFlavor::Msvc(..), LinkerFlavorCli::Msvc(..))
+                | (LinkerFlavor::EmCc, LinkerFlavorCli::EmCc)
+                | (LinkerFlavor::Bpf, LinkerFlavorCli::Bpf)
+                | (LinkerFlavor::Ptx, LinkerFlavorCli::Ptx) => return true,
+                _ => {}
+            }
+
+            // 2. or, the flavor is legacy and survives this roundtrip.
+            cli == self.with_cli_hints(cli).to_cli()
```

Comment:
@petrochenkov I'm not sure about `check_compatibility` here: my thinking was to consider `to_cli` as if it were called `to_legacy_cli`, and keep the old lossy conversions. The new flavors would only need to look at the principal flavor for compatibility. Do you think that is correct ?

The target consistency tests still pass, expectedly since the conversion behavior shouldn't have changed either: no new flavors are used yet. 

Using the flavor names you described before is of course compatible here, and `-Clink-self-contained=+linker -Clinker-flavor=gnu-lld-cc -Zunstable-options` works in the end to end test for example.

Reply:
The new logic looks correct to me.

Although we need to eventually update `to_cli` to not mean `to_legacy_cli`.
`to_cli` is used to produce correct json, and with the new flavor supported it's still possible (and easier) to get correct json without resorting to legacy flavors.